monai/transforms/io/array.py

```diff
@@ -263,7 +263,14 @@ def __call__(self, filename: Union[Sequence[PathLike], PathLike], reader: Option
             )
 
         img_array: NdarrayOrTensor
-        img_array, meta_data = reader.get_data(img)
+        img_array, meta_data = reader.get_data(img_obj)
+        # Close image objects if they have close attribute
```

Comment:
this could be done in the reader's `get_data()` implementation? also I'm not sure if the readers want to close objects at every iteration

Reply:
@wyli, I see your point but implementing in `get_data` will reduce the flexibility of the reader component. However, here, if you look at line 239, the image object is being read at each call, so closing it at the end of the call should be fine since we are not returning the image object. Do you agree?

Reply:
I don't think within this generic LoadImage should understand any attributes of `img_obj`

Reply:
Then do you suggest to leave the warning as is and close the issue?

Reply:
not sure which reader is creating this obj, if it's in the WSIreader, perhaps close it right after this line? https://github.com/Project-MONAI/MONAI/blob/9efc6e9a966751a91f8eaac36e2d205f40686b6c/monai/data/image_reader.py#L1346

Reply:
It is only happening for TiffFile backend. Do you agree that for reading an image (WSI, PNG, PIL, etc.) a user might want to create an image object and call `get_data` multiple times? This is currently a supported behavior and I believe we should not change that.

On the other hand, modifying `LoadImage` to close the image object at the end of the call [if possible] does not change ANY behavior of `LoadImage`, does it?

cc @Nic-Ma @ericspod

Reply:
yeah, the current PR breaks the LoadImage API...decoupling the backend objects from LoadImage call is important, maybe you can create some other APIs for wsireader users who want to get_data but not close the object if there are feature requests

Reply:
or if it's just a warning maybe we can ignore it... I just found the warning message when I was looking at something else, there's no other issues at the moment

Reply:
Maybe I am missing something but LoadImage API accepts the image reader as input and not the image object, so the object is being created in the call and then abandoned when the call is over. Then for the next time, a new object needs to be created again, right?

If so, the object is not being re-used and closing the object would be actually a desired behavior. Am I missing something here?

Reply:
the idea is that the image reader should define the backend object details. so if there's really a common requirement among the backends to postprocess the objects, the LoadImage should have some command like `reader.finalize(img_obj)`

Reply:
That I agree! It would be nice to have a method common to all readers inherited from `ImageReader` so that it closes the image object if possible. So do you suggest to add such method to `ImageReader` abstract class?

Reply:
I don't think so, because the other 6 reader implementations don't need this interface. I'd suggest an update in wsireader with the additional arg `wsireader.get_data(...keep_obj=False)`, so the user specifically set `keep_obj=True` the object will be left open after this call

Reply:
It is not a requirement for all WSI reader (only TiffFile). Also that breaks the abstraction as they are inherited `ImageReader`.

I believe the solution in this PR is non-intrusive and does not change any API or intended behavior of ANY component in MONAI as I described above. The other proposed solution is breaking abstraction, and causing backward incompatibility for other use cases (for instance in `PatchWSIReader` where we are caching image objects).

@Nic-Ma @ericspod any thought here?

Reply:
for my suggested solution, you can change the call here to use `keep_obj=True`: https://github.com/Project-MONAI/MONAI/blob/27362e482761985daac8d5ef732eb57b50c741f1/monai/data/wsi_datasets.py#L146

it's non-breaking because  `PatchWSIReader._get_data` is a private method

Reply:
Hi @drbeh @wyli ,

I think the idea of adding param in `get_data()` is OK, but the arg name `keep_obj=True` may be not clear enough.

Thanks.

Reply:
I don't feel this is really the correct solution. `LoadImage` isn't concerned with how the readers read and present data, so details like file closing or other finalisation is left up to the reader side. Adding something to close objects in `LoadImage` spreads concern about file handling between these two classes. If we can change the concerned readers to have an added constructor/method argument to control whether it's methods should close files or not then this code wouldn't needed here.

Reply:
Hi @drbeh , @ericspod , @wyli ,

If I understand the problem correctly, how about this: add some parameter to the `WSIReader` init function to control whether we should close the file after reading. Then we can pass additional args to the reader in `LoadImage` by the `arg` or `kwargs`: https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/io/array.py#L143
What do you think?

Thanks.

Reply:
@Nic-Ma I agree that solution would work but it make the `WSIReader` API unnecessarily crowded, considering that in WSI images, we usually don't want to close the image object  and also we rarely use LoadImage (except for loading in a low resolution level like MIL example).

If everyone agrees, I would not change anything here and let the warning of TiffFile be there in the tests.
cc @Nic-Ma @wyli @ericspod